internal/cortex/frontend/transport/handler.go

```diff
@@ -165,6 +220,69 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func (f *Handler) updateFailedQueryCache(err error, queryExpressionNormalized string, queryExpressionRangeLength int, r *http.Request) {
+	// Extracting error code from error string.
+	codeExtract := f.errorExtract.FindStringSubmatch(err.Error())
+
+	// Checking if error code extracted successfully.
+	if codeExtract == nil || len(codeExtract) < 2 {
+		level.Error(util_log.WithContext(r.Context(), f.log)).Log(
+			"msg", "Error string regex conversion error")
+		return
+	}
+
+	// Converting error code to int.
+	errCode, strConvError := strconv.Atoi(codeExtract[1])
+
+	// Checking if error code extracted properly from string.
+	if strConvError != nil {
+		level.Error(util_log.WithContext(r.Context(), f.log)).Log(
+			"msg", "String to int conversion error")
+		return
+	}
+
+	// If error should be cached, store it in cache.
+	if !isCacheableError(errCode) {
+		level.Debug(util_log.WithContext(r.Context(), f.log)).Log(
+			"msg", "Query not cached due to non-cacheable error code")
+		return
+	}
+
+	// Checks if queryExpression is already in cache, and updates time range length value to min of stored and new value.
+	if contains, _ := f.lruCache.ContainsOrAdd(queryExpressionNormalized, queryExpressionRangeLength); contains {
+		if oldValue, ok := f.lruCache.Get(queryExpressionNormalized); ok {
+			queryExpressionRangeLength = min(queryExpressionRangeLength, oldValue.(int))
+		}
+		f.lruCache.Add(queryExpressionNormalized, queryExpressionRangeLength)
+	}
+
+	level.Debug(util_log.WithContext(r.Context(), f.log)).Log(
+		"msg", "Query cached", "response ")
```

Comment:
The `"response "` part is not right. Please use the following log line and paste the long lines from your testing in the PR description.
```
...Log(
    "msg", "Cached a failed query",
    "normaized_query", queryExpressionNormalized,
    "range_seconds", queryExpressionRangeLength,
    "error", err,
)
```